typing/typedecl_variance.ml

```diff
@@ -188,14 +202,35 @@ let compute_variance_type env ~check (required, loc) decl tyl =
           !tvl2 null in
       Btype.backtrack snap;
       let (c1,n1) = get_upper v1 and (c2,n2,i2) = get_lower v2 in
-      if c1 && not c2 || n1 && not n2 then
-        if List.exists (eq_type ty) fvl then
-          let code = if not i2 then No_variable
-                     else if c2 || n2 then Variance_not_reflected
-                     else Variance_not_deducible in
-          raise (Error (loc, Bad_variance (code, (c1,n1,false), (c2,n2,false))))
-        else
-          Btype.iter_type_expr check ty
+      if c1 && not c2 || n1 && not n2 then begin
+        match List.find_opt (eq_type ty) fvl with
+        | Some variable ->
+            let code =
+              if not i2 then
+                Variance_variable_error
+                  { error = No_variable
+                  ; context
+                  ; variable
+                  }
+              else if c2 || n2 then
+                Variance_variable_error
+                  { error = Variance_not_reflected
+                  ; context
+                  ; variable
+                  }
+              else
+                Variance_variable_error
+                  { error = Variance_not_deducible
+                  ; context
+                  ; variable
+                  }
+            in
```

Comment:
Two remarks:
- `code` is a cryptic variable name, I would use `variance_error`
- there is a lot of visual redudancy in the record constructions, why not push the conditionals inside a `let error = if ... in` declaration and then have a single record construction?